iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -526,6 +526,68 @@ LogicalResult SortOp::generateScalarImplementation(OpBuilder &b, Location loc,
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// FftOp
+//===----------------------------------------------------------------------===//
+
+static LogicalResult verifyFftOp(FftOp op) {
+  auto length = op.getFftLength();
+  if (length == ShapedType::kDynamicSize) return failure();
+  // Expect to be a power of 2.
+  if (length & (length - 1)) return failure();
+  if (!op.getNumInputs() || !op.isScalar(op.getInputOperand(0))) {
+    return op.emitOpError("expected to carry `stage` input");
+  }
+  if (op.getNumOutputs() != 2) {
+    return op.emitOpError("expected outputs to be real and imag tensor/memref");
+  }
+  return success();
+}
+
+void FftOp::getEffects(
+    SmallVectorImpl<SideEffects::EffectInstance<MemoryEffects::Effect>>
+        &effects) {
+  SmallVector<Value> inputBuffers = getInputBufferOperands();
+  SmallVector<Value> outputBuffers = getOutputBufferOperands();
+  getEffectsImpl(effects, getOperation()->getResults(), inputBuffers,
+                 outputBuffers);
+}
+
+SmallVector<StringRef> FftOp::getLoopIteratorTypes() {
+  // There are `rank-1` outer loops. The fft itselfs has two loops, one is for
+  // the stage, and another is merge step. Thus, there are `rank+1` in total.
+  SmallVector<StringRef> iteratorTypes(getOperandRank() + 1,
```

Comment:
Hmm, not sure I follow. THe number of loops here should refer to the specific stage and not for the whole FFT computation.

Reply:
After having some prototype of `FftOp::generateScalarImplementation`, I think we still need the extra loop. Otherwise we will create subview + linalg.generic op like old implementation of scatter op. E.g.,

```mlir
LogicalResult FftOp::generateScalarImplementation(OpBuilder &b, Location loc,
                                                  ValueRange ivs) {
  Value real = getReal();
  Value imag = getImag();
  Value stage = getStage();
  Value one = b.create<ConstantIndexOp>(loc, 1);
  Value entireSize = b.create<ShiftLeftOp>(loc, one, stage);
  Value halfSize = b.create<SignedShiftRightOp>(loc, stage, one);

  SmallVector<Value> ones(getOperandRank(), one);
  SmallVector<Value> sizes(getOperandRank(), one);
  sizes.back() = halfSize;
  Value lhsReal =
      b.create<memref::SubViewOp>(loc, real, ivs, sizes, ones);
  Value lhsImag =
      b.create<memref::SubViewOp>(loc, imag, ivs, sizes, ones);

  SmallVector rhsIvs(ivs.begin(), ivs.end());
  rhsIvs.back() = b.create<AddIOp>(loc, rhsIvs.back(), halfSize);
  Value rhsReal =
      b.create<memref::SubViewOp>(loc, real, rhsIvs, sizes, ones);
  Value rhsImag =
      b.create<memref::SubViewOp>(loc, imag, rhsIvs, sizes, ones);

   // create a generic op which takes (lhs, rhs).
}
```

With the extra loop, we can still tile and distribute it. And we will create `memref::LoadOp` in the scalar implementation. The range of the extra loop is from 0 to halfSize, and the step is one in this case.

Reply:
Actually having the `linalg.generic` in there is probably right. Lets quickly chat on this over GVC?